src/part_file.cpp

```diff
@@ -210,24 +213,24 @@ namespace libtorrent
 			return -1;
 		}
 
-		int slot = i->second;
+		int const slot = i->second;
 
 		open_file(file::read_write | file::attribute_hidden, ec);
 		if (ec) return -1;
 
 		l.unlock();
 
-		boost::int64_t slot_offset = boost::int64_t(m_header_size) + boost::int64_t(slot) * m_piece_size;
+		boost::int64_t const slot_offset = calc_offset(slot, m_header_size, m_piece_size);
 		return m_file.readv(slot_offset + offset, bufs, num_bufs, ec);
 	}
 
 	void part_file::open_file(int mode, error_code& ec)
 	{
-		if (m_file.is_open()
-			&& ((m_file.open_mode() & file::rw_mask) == mode
-				|| mode == file::read_only)) return;
+		if (m_file.is_open() && (mode == file::read_only
+			|| (m_file.open_mode() & file::rw_mask) == (mode & file::rw_mask)))
```

Comment:
I think the original intention for that expression was to say: `(m_file.open_mode() & mode) == mode`, i.e. is every bit in `mode` is set in `m_file.open_mode()`? If so, we have at least the capabilities we need.

Reply:
I think that expression has the same behavior as yours, but I'd say is a bit simpler (assuming you expect the mode to be a bitmask)

Reply:
ah, I suppose it would have to look like this:
`(m_file.open_mode() & mode & file::rw_mask) == (mode & file::rw_mask)`

Reply:
Now I'm starting to hesitate if any of given expressions are correct for this function.
Maybe we should list constraints that come from possible use cases?

Reply:
well, in practice I don't think that's important because libtorrent doesn't actually use that mode. In fact, in a future branch (not master, but `new-disk-io`) that mode has been removed

Reply:
Next let's define compatibility of different file modes in terms of `part_file` operations.

Reply:
Does it make sense to consider any of these values: `sparse`, `no_atime`, `random_access`, `lock_file`, `no_cache`?
I didn't include file attributes, because their modification doesn't require file handle.

Reply:
I don't think so. I would imagine `random_access` would always be set (but maybe it isn't). Otherwise I don't think really any of those flags are critical.

Reply:
I would think it's more reasonable to ignore `random_access` and just make sure the read-write permissions are compatible.

what I think you want to check is to make sure that all bits set in `mode` are also set in `m_file.open_mode()`

Reply:
So my initial change in `part_file::open_file` is correct?